### PR TITLE
imagemagick: include configure.xml (#353169)

### LINKED ISSUE #82749: imagemagick: unable to access configure file `configure.xml'
**Describe the bug**
I would like to use a shell script which would use the output `convert -list configure`, however on my nixos, along with some useful text, it prints the error:
`convert: unable to access configure file 'configure.xml' @ warning/configure.c/GetConfigureOptions/677.`

**To Reproduce**
1. Run `convert -list configure`
2. Look at the output

**Expected behavior**
No error, and some additional values are printed from the xml (as I cannot produce the correct output, I have no idea how it would look like).

**Additional context**
I'm trying to run one of the Fred's ImageMagick scripts which can be located at http://www.fmwconcepts.com/imagemagick/textcleaner/index.php.

**Metadata**
Please run `nix-shell -p nix-info --run "nix-info -m"` and paste the result.

these paths will be fetched (0.05 MiB download, 0.28 MiB unpacked):
  /nix/store/5h35s50503spks8gkxdv3kk4szz8vx9g-bash-interactive-4.4-p23-dev
copying path '/nix/store/5h35s50503spks8gkxdv3kk4szz8vx9g-bash-interactive-4.4-p23-dev' from 'https://cache.nixos.org'...
 - system: `"x86_64-linux"`
 - host os: `Linux 4.19.109, NixOS, 19.09.2260.3b9b10e49be (Loris)`
 - multi-user?: `yes`
 - sandbox: `yes`
 - version: `nix-env (Nix) 2.3.3`
 - channels(root): `"nixos-19.09.2260.3b9b10e49be, nixos-unstable-20.09pre216896.0729b8c55e0"`
 - nixpkgs: `/nix/var/nix/profiles/per-user/root/channels/nixos`

Maintainer information:
```yaml
# a list of nixpkgs attributes affected by the problem
attribute: imagemagick
# a list of nixos modules affected by the problem
module:
```

### PR DESCRIPTION
**Pull request by akgrant43**
`magick -list configure` relies on the configuration file being in a known location.  Copy `configure.xml` to `share/ImageMagick-<version>/`

fixes https://github.com/NixOS/nixpkgs/issues/82749

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#linking-nixos-module-tests-to-a-package) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) (or backporting [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) and [24.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

**Review by dotlambda — CHANGES_REQUESTED** [review]
`magick -list configure` produces outputs like
```
CFLAGS                -I/usr/include/libxml2    -I/usr/include/libraw  -I/usr/include/libpng16   -I/usr/include/pango-1.0 -I/usr/include/glib-2.0 -I/usr/lib64/glib-2.0/include -I/usr/include/cairo -I/usr/include/pixman-1 -I/usr/include/freetype2 -I/usr/include/libpng16 -I/usr/include/libdrm -I/usr/include/harfbuzz  -I/usr/include/OpenEXR -pthread -I/usr/include/libdrm   -I/usr/include/openjpeg-2.2    -I/usr/include/freetype2 -I/usr/include/libpng16  -I/usr/include/freetype2 -I/usr/include/libpng16  -pthread     -fopenmp -g -O2 -Wall -mtune=sandybridge -fexceptions -pthread -DMAGICKCORE_HDRI_ENABLE=1 -DMAGICKCORE_QUANTUM_DEPTH=16
```
which is definitely not correct

**Comment by rhendric** [comment]
I think even if those FHS paths are replaced with their Nix equivalents, this is still a bad idea because it would establish a dependency from the `out` output on inputs' header files and such. `imagemagick` has a separate `dev` output for those dependencies.

**Comment by akgrant43** [comment]
My motivation for looking at this is that I have some 3rd-party programs which rely on the output of `magick -list configure`, as far as I know, they're not relying on the paths.

I haven't come across derivations with multiple outputs before, but I assume your concern is because the other outputs can be removed, e.g. garbage collected, which will of course break the reported path?

What if we remove the paths from the report?  That isn't ideal as it is missing information which is available on other platforms, but is better than the current situation where no information is included.

The output in that case would be something like:

```
$ magick -list configure

Path: /nix/store/wv41k579w2vpyf01smr1dw21k5ii0127-imagemagick-7.1.1-39/share/ImageMagick-7/configure.xml

Name                  Value
-------------------------------------------------------------------------------
CC                    gcc
CHANNEL_MASK_DEPTH    32
COPYRIGHT             Copyright (C) 1999 ImageMagick Studio LLC
CXX                   g++
CXXFLAGS              -g -O2 -pthread
DEFS                  -DHAVE_CONFIG_H
DELEGATES             bzlib cairo djvu fftw fontconfig freetype heic jng jp2 jpeg jxl lcms lqr lzma openexr pangocairo png raw rsvg tiff webp x xml zlib zstd
DISTCHECK_CONFIG_FLAGS 'CC=gcc' 'CXX=g++'  --disable-deprecated  --with-quantum-depth=16  --with-jemalloc=no  --with-umem=no  --with-autotrace=no  --with-dps=no  --with-flif=no  --with-fpx=no  --with-gslib=no  --with-fontpath=  --with-gs-font-dir=no  --with-uhdr=no  --with-wmf=no  --with-perl=no 
FEATURES              Channel-masks(32-bit) Cipher DPC HDRI OpenMP
GIT_REVISION          21316
HOST                  x86_64-pc-linux-gnu
LIB_VERSION           0x711
LIB_VERSION_NUMBER    7,1,1,39
NAME                  ImageMagick
PCFLAGS               -fopenmp -DMAGICKCORE_HDRI_ENABLE=1 -DMAGICKCORE_QUANTUM_DEPTH=16 -DMAGICKCORE_CHANNEL_MASK_DEPTH=32
QuantumDepth          16
RELEASE_DATE          2024-10-02
SECURITY_POLICY       open
TARGET_CPU            x86_64
TARGET_OS             linux-gnu
TARGET_VENDOR         pc
VERSION               7.1.1
WEBSITE               https://imagemagick.org

Path: [built-in]

Name                  Value
-------------------------------------------------------------------------------
DELEGATES             bzlib cairo djvu fftw fontconfig freetype heic jng jp2 jpeg jxl lcms lqr lzma openexr pangocairo png raw rsvg tiff webp x xml zlib zstd
FEATURES              Cipher DPC HDRI OpenMP(4.5) 
MAGICK_TEMPORARY_PATH /tmp
NAME                  ImageMagick
QuantumDepth          Q16
```

**Comment by dotlambda** [comment]
> I assume your concern is because the other outputs can be removed, e.g. garbage collected

No, `dev` can't he garbage collected if `out` references it, unless the latter is garbage collected as well.
But the real problem is that installing imagemagick should be possible without downloading `dev`.

> What if we remove the paths from the report?

Even so, there is no reason to assume the remaining information is correct for our build. In particular, I assume the output should be different for the small and big variants.

**Comment by akgrant43** [comment]
> Even so, there is no reason to assume the remaining information is correct for our build. In particular, I assume the output should be different for the small and big variants.

The `configure.xml` file is generated by the `./configure` at build time.  Here's the diff for the normal and big variants of imagemagick:

```
$ diff configure.xml big-configure.xml
<   <configure name="DELEGATES" value="bzlib cairo djvu fftw fontconfig freetype heic jng jp2 jpeg jxl lcms lqr lzma openexr pangocairo png raw rsvg tiff webp x xml zlib zstd"/>
<   <configure name="DISTCHECK_CONFIG_FLAGS" value="'CC=gcc' 'CXX=g++'  --disable-deprecated  --with-quantum-depth=16  --with-jemalloc=no  --with-umem=no  --with-autotrace=no  --with-dps=no  --with-flif=no  --with-fpx=no  --with-gslib=no  --with-fontpath=  --with-gs-font-dir=no  --with-uhdr=no  --with-wmf=no  --with-perl=no "/>
---
>   <configure name="DELEGATES" value="bzlib cairo djvu fftw fontconfig freetype gslib heic jng jp2 jpeg jxl lcms lqr lzma openexr pangocairo png ps raw rsvg tiff webp x xml zlib zstd"/>
```

I don't know why the distcheck flags are left out of the big variant, but you can see that `gslib` has been added to the delegates in the big variant.

**Comment by akgrant43** [comment]
I should add that I was including the wrong version of configure.xml previously.  There's a copy that's included with the source, and one that is generated by `./configure`.  I haven't committed the code yet, but I'm now using the locally generated one (and stripping out any paths as discussed above).

File: pkgs/applications/graphics/ImageMagick/default.nix [hunk 1]
```diff
@@ -122,7 +122,7 @@ stdenv.mkDerivation (finalAttrs: {
     (cd "$dev/include" && ln -s ImageMagick* ImageMagick)
     moveToOutput "bin/*-config" "$dev"
     moveToOutput "lib/ImageMagick-*/config-Q16HDRI" "$dev" # includes configure params
-    cp $src/www/source/configure.xml $out/share/ImageMagick-*/
+    grep -v '/nix/store' $dev/lib/ImageMagick-*/config-Q16HDRI/configure.xml > $out/share/ImageMagick-7/configure.xml
```
Comment (dotlambda) [inline pkgs/applications/graphics/ImageMagick/default.nix:125]: Are you sure that path stays the same?

**Review by dotlambda — COMMENTED** [review]
Squash.

File: pkgs/applications/graphics/ImageMagick/default.nix (hunk 1 above)
Comment (akgrant43) [reply pkgs/applications/graphics/ImageMagick/default.nix]: Thanks, I've corrected this in the squashed commit.

**Comment by akgrant43** [comment]
> Squash.

Done, thanks.

File: pkgs/applications/graphics/ImageMagick/default.nix (hunk 1 above)
Comment (dotlambda) [reply pkgs/applications/graphics/ImageMagick/default.nix]: I was also referring to `Q16HDRI`.

Comment (akgrant43) [reply pkgs/applications/graphics/ImageMagick/default.nix]: That was hardcoded in the derivation prior to my PR (the `moveToOutput` line just prior to my addition).

Doing a quick search through the source code it doesn't seem to be set anywhere, although it is used in multiple places.

Comment (bloxx12) [reply pkgs/applications/graphics/ImageMagick/default.nix]: > Are you sure that path stays the same?

It is hard-coded into several files and hasn't changed in several years. 
I think it'd be a good idea to comment this part of the code briefly though.

**Comment by akgrant43** [comment]
I've made a mistake and included a change from an unrelated commit in this PR.  I'll close the PR and resubmit it.
My apologies for the inconvenience.

**Comment by akgrant43** [comment]
Resubmitted in: https://github.com/NixOS/nixpkgs/pull/353842

**Comment by dotlambda** [comment]
You should just have force pushed instead.

**Comment by akgrant43** [comment]
> You should just have force pushed instead.

It was a force push that caused the problem, and I'm not sure exactly what I did wrong.  I think it was because I committed the change locally, and then did a `rebase -i HEAD~3` instead of `HEAD~2` (and then force pushed).  I'm afraid I don't know how to undo that mistake.

**Comment by akgrant43** [comment]
> > I'm afraid I don't know how to undo that mistake.
> 
> `git rebase -i HEAD~2` and remove the offending commit

Thanks for the explanation!  I've removed the offending commit and force-pushed.  The repository shows the branch as being one commit ahead of NixOS/nixpkgs/master, but this PR hasn't updated, presumably because it is closed (and I can't reopen it).



https://github.com/NixOS/nixpkgs/pull/353842 has the same changes.

**Pull request closed without merging**